### PR TITLE
Replace meta_tags.csv with hparams.yaml

### LINKED ISSUE #1212: Add support for loading flattened meta_tags.csv
### Motivation
PL+TensorBoard can log hierarchical dict after #1152, however, `meta_tags.csv` has been disabled by the change.

### Pitch
- Make `meta_tags.csv` back to a hierarchical dict based on their delimiter.

### Alternatives
62de7948634b1cd8c7e494f4cd9a02625cd3f602

### Additional context
1. We can consider the deprecation of `meta_tags.csv`, then adopt `config.yaml`.
1. We can interpret primitive-type parameters through the files, so it is needed to rethink parameter sanitization or update docs.

### PR DESCRIPTION
**Pull request by S-aiueo32**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #1212 .

### CODE REVIEW

File: pytorch_lightning/core/saving.py [hunk 1]
```diff
@@ -1,5 +1,8 @@
+import ast
+import collections
```
Comment (awaelchli) [inline pytorch_lightning/core/saving.py]: is this import needed? I can't see it.

File: pytorch_lightning/loggers/tensorboard.py [hunk 2]
```diff
@@ -37,7 +38,6 @@ class TensorBoardLogger(LightningLoggerBase):
         \**kwargs: Other arguments are passed directly to the :class:`SummaryWriter` constructor.
 
     """
-    NAME_CSV_TAGS = 'meta_tags.csv'
```
Comment (awaelchli) [inline pytorch_lightning/loggers/tensorboard.py]: I think it would be good to have it as a constant somewhere instead of hard-coding the string.

File: requirements.txt [hunk 3]
```diff
@@ -5,4 +5,4 @@ numpy>=1.16.4
 torch>=1.1
 tensorboard>=1.14
 future>=0.17.1  # required for builtins in setup.py
-
+pyyaml>=5.3.1
```
Comment (Borda) [inline requirements.txt]: is it really the minimal version we can use?

File: pytorch_lightning/loggers/tensorboard.py (hunk 2 above)
Comment (S-aiueo32) [reply pytorch_lightning/loggers/tensorboard.py]: done aacc074983c67837c52ab2de65143e8d6285103a

File: pytorch_lightning/core/saving.py (hunk 1 above)
Comment (S-aiueo32) [reply pytorch_lightning/core/saving.py]: done d922f2f485c10e177ad92ac9d9a64a12085a497b

File: requirements.txt (hunk 3 above)
Comment (S-aiueo32) [reply requirements.txt]: it's the current latest version

Comment (Borda) [reply requirements.txt]: Could we put there like 3.x or lover since we are using quite general features?

Comment (awaelchli) [reply requirements.txt:8]: @S-aiueo32 Don't forget to also add it to the environment.yml file

**Comment by Borda** [comment]
@S-aiueo32 how is it going? it would be great to have it in next week release 0.7.2

**Comment by S-aiueo32** [comment]
@Borda I'm sorry, but I'm too busy for about two weeks this week to be able to respond. I recognize that the major changes are complete and all that remains is testing and minor revisions. If you don't mind, could you let the core contributor take over the work?

**Comment by Borda** [comment]
changelog need to be rebased on new release #1419

**Comment by Borda** [comment]
@S-aiueo32 may you pls allow editing your PR, I have prepared to rebase, but don't have permission to push it... :robot:

**Comment by S-aiueo32** [comment]
@Borda confirmed?

**Comment by Borda** [comment]
@S-aiueo32 It shall be an option in this PR, kind of check box...
https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork

**Comment by S-aiueo32** [comment]
@Borda 🤔 
スクリーンショット 2020-04-12 17 59 19

**Comment by Borda** [comment]
Cool, I will try later today again =)

**Comment by Borda** [comment]
still getting:
```
remote: Permission to S-aiueo32/pytorch-lightning.git denied to Borda.
fatal: unable to access 'https://github.com/S-aiueo32/pytorch-lightning.git/': The requested URL returned error: 403
```

**Comment by S-aiueo32** [comment]
@Borda  hmm... 🤔 It seems that the button doesn't work.
I added you to the collaborator of S-aiueo32/pytorch-lightning. Please check it out because the request is pending now.

**Comment by Borda** [comment]
@S-aiueo32 it seems that it is not implemented in all loggers, mind fixing it or I will do it..
for that I would need to take you changes and make new PR... :rabbit:

**Comment by S-aiueo32** [comment]
@Borda
Currently, other loggers do not implement meta_tags, but do you think that support should be extended? I'm cautious about extending that, as any variable can now be stored in a checkpoint as hparams.

Anyway, it looks like I'll be able to resume my contributions this weekend, so feel free to let me do the work for you.

**Comment by Borda** [comment]
Well, we would like to make TensorBoard as default logger for testing #1478 
it is true we not all longer may support hparams logging, so saving them in yaml next to checkpoint make sense but then we need to be careful about using the same folder for multiple experiments and overwriting this yaml file...

@PyTorchLightning/core-contributors where shall be the Hparams saved? logger or checkpoint?

**Comment by Borda** [comment]
@S-aiueo32 how is it going?

**Comment by Borda** [comment]
@S-aiueo32 how is it going?

**Comment by S-aiueo32** [comment]
@Borda 
I'm sorry for my late response. I've fixed for test logger changes and all related test are passed.
Now CI checks are failing because of runtime error around horovod.

File: pytorch_lightning/core/lightning.py [hunk 4]
```diff
@@ -1519,10 +1521,20 @@ def __init__(self, hparams):
...
+            warnings.warn('tags_csv argument is deprecated, use hparams_file argument instead.')
+
+        if hparams_file is not None:
+            extension = hparams_file.split('.')[-1]
+            if extension.lower() in ['csv']:
+                hparams = load_hparams_from_tags_csv(hparams_file)
+            elif extension.lower() in ['yml', 'yaml']:
+                hparams = load_hparams_from_yaml(hparams_file)
+            else:
+                raise ValueError('.csv, .yml or .yaml is required for hparams_file')
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1534]:
```suggestion
                raise ValueError('.csv, .yml or .yaml is required for `hparams_file`')
```

File: pytorch_lightning/core/lightning.py [hunk 5]
```diff
@@ -1519,10 +1521,20 @@ def __init__(self, hparams):
...
-            # add the hparams from csv file to checkpoint
-            hparams = load_hparams_from_tags_csv(tags_csv)
-            hparams.__setattr__('on_gpu', False)
-            checkpoint['hparams'] = vars(hparams)
+            hparams_file = tags_csv
+            warnings.warn('tags_csv argument is deprecated, use hparams_file argument instead.')
+
+        if hparams_file is not None:
+            extension = hparams_file.split('.')[-1]
+            if extension.lower() in ['csv']:
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1529]:
```suggestion
            if extension.lower() in ('csv'):
```

File: pytorch_lightning/core/lightning.py [hunk 6]
```diff
@@ -1519,10 +1521,20 @@ def __init__(self, hparams):
...
+                hparams = load_hparams_from_tags_csv(hparams_file)
+            elif extension.lower() in ['yml', 'yaml']:
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1531]:
```suggestion
            elif extension.lower() in ('yml', 'yaml'):
```

File: pytorch_lightning/core/lightning.py [hunk 7]
```diff
@@ -1519,10 +1521,20 @@ def __init__(self, hparams):
             checkpoint = torch.load(checkpoint_path, map_location=lambda storage, loc: storage)
 
         if tags_csv is not None:
-            # add the hparams from csv file to checkpoint
-            hparams = load_hparams_from_tags_csv(tags_csv)
-            hparams.__setattr__('on_gpu', False)
-            checkpoint['hparams'] = vars(hparams)
+            hparams_file = tags_csv
+            warnings.warn('tags_csv argument is deprecated, use hparams_file argument instead.')
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1525]: uze `zero_rand_warn` and specify from when it is deprecated and till when
```suggestion
            rank_zero_warn('`tags_csv` argument is deprecated in v0.7.6. Will be removed v0.9.0', DeprecationWarning)
```

File: pytorch_lightning/core/lightning.py [hunk 8]
```diff
@@ -1436,6 +1437,7 @@ def load_from_checkpoint(
             checkpoint_path: str,
             map_location: Optional[Union[Dict[str, str], str, torch.device, int, Callable]] = None,
             tags_csv: Optional[str] = None,
+            hparams_file: Optional[str] = None,
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1440]:
```suggestion
            hparams_file: Optional[str] = None,
            tags_csv: Optional[str] = None,  # backward compatible, todo: remove in v0.9.0
```

File: pytorch_lightning/core/lightning.py [hunk 9]
```diff
@@ -1534,15 +1546,18 @@ def _load_model_state(cls, checkpoint: Dict[str, Any], *args, **kwargs) -> 'Ligh
 
         if cls_takes_hparams:
             if ckpt_hparams is not None:
-                is_namespace = checkpoint.get('hparams_type', 'namespace') == 'namespace'
-                hparams = Namespace(**ckpt_hparams) if is_namespace else ckpt_hparams
+                hparams_type = checkpoint.get('hparams_type')
```
Comment (Borda) [inline pytorch_lightning/core/lightning.py:1549]:
```suggestion
                hparams_type = checkpoint.get('hparams_type', 'namespace')
```
use some default...

File: pytorch_lightning/core/saving.py [hunk 10]
```diff
@@ -48,30 +51,32 @@ def on_hpc_load(self, checkpoint: Dict[str, Any]) -> None:
...
         tags = {row[0]: convert(row[1]) for row in list(csv_reader)[1:]}
-    ns = Namespace(**tags)
-    return ns
+
+    return tags
+
+
+def load_hparams_from_yaml(config_yaml: str) -> Dict[str, Any]:
+    if not os.path.isfile(config_yaml):
+        log.warning(f'Missing Tags: {config_yaml}.')
```
Comment (Borda) [inline pytorch_lightning/core/saving.py:68]:
```suggestion
        rank_zero_warn(f'Missing Tags: {config_yaml}.', RuntimeWarning)
```

File: pytorch_lightning/core/saving.py [hunk 11]
```diff
@@ -48,30 +51,32 @@ def on_hpc_load(self, checkpoint: Dict[str, Any]) -> None:
...
-    for c in constructors:
-        try:
-            return c(val)
-        except ValueError:
-            pass
-    return val
+    try:
+        return ast.literal_eval(val)
+    except (ValueError, SyntaxError) as e:
+        logging.debug(e)
```
Comment (Borda) [inline pytorch_lightning/core/saving.py:81]: use `from pytorch_lightning import _logger as log`

File: pytorch_lightning/loggers/tensorboard.py [hunk 12]
```diff
@@ -147,15 +150,11 @@ def save(self) -> None:
...
 
         # save the metatags file
-        with open(meta_tags_path, 'w', newline='') as csvfile:
-            fieldnames = ['key', 'value']
-            writer = csv.DictWriter(csvfile, fieldnames=fieldnames)
-            writer.writerow({'key': 'key', 'value': 'value'})
-            for k, v in self.tags.items():
-                writer.writerow({'key': k, 'value': v})
+        with open(hparams_file, 'w', newline='') as f:
+            yaml.dump(self.hparams, f)
```
Comment (Borda) [inline pytorch_lightning/loggers/tensorboard.py:157]: I would encapsulate this to a function next to the load to be clear it is symmetric and add doctest with saving and load are equal...

File: pytorch_lightning/trainer/training_io.py [hunk 13]
```diff
@@ -344,9 +344,16 @@ def dump_checkpoint(self):
...
-            checkpoint['hparams_type'] = 'namespace' if is_namespace else 'dict'
+            checkpoint['hparams_type'] = model.hparams.__class__.__name__
+            if checkpoint['hparams_type'] == 'dict':
+                checkpoint['hparams'] = model.hparams
+            elif checkpoint['hparams_type'] == 'Namespace':
+                checkpoint['hparams'] = vars(model.hparams)
+            else:
+                raise ValueError(
+                    'The acceptable hparams type is dict or argparse.Namespace,',
+                    f'not {checkpoint["hparams_type"]}'
```
Comment (Borda) [inline pytorch_lightning/trainer/training_io.py:355]:
```suggestion
                    f' not {checkpoint["hparams_type"]}'
```

File: tests/trainer/test_trainer.py [hunk 14]
```diff
@@ -193,12 +198,43 @@ def test_loading_meta_tags(tmpdir):
...
+        hparams = yaml.load(f)
+
+    # save as legacy meta_tags.csv
     tags_path = os.path.join(path_expt_dir, 'meta_tags.csv')
+    with open(tags_path, 'w') as f:
+        fieldnames = ['key', 'value']
+        writer = csv.DictWriter(f, fieldnames=fieldnames)
+        writer.writerow({'key': 'key', 'value': 'value'})
+        for k, v in hparams.items():
+            writer.writerow({'key': k, 'value': v})
```
Comment (Borda) [inline tests/trainer/test_trainer.py:214]: move this to a func next to save to YAML

File: tests/trainer/test_trainer.py [hunk 15]
```diff
@@ -193,12 +198,43 @@ def test_loading_meta_tags(tmpdir):
     logger.log_hyperparams(hparams)
     logger.save()
 
-    # load tags
+    # load hparams
     path_expt_dir = tutils.get_data_path(logger, path_dir=tmpdir)
+    hparams_path = os.path.join(path_expt_dir, 'hparams.yaml')
+    with open(hparams_path, 'r') as f:  # store as a dict
+        hparams = yaml.load(f)
```
Comment (Borda) [inline tests/trainer/test_trainer.py:205]: this shall use loading func

File: tests/trainer/test_trainer.py [hunk 16]
```diff
@@ -193,12 +198,37 @@ def test_loading_meta_tags(tmpdir):
...
+    hparams = load_hparans_from_yaml(hparams_path)
+
+    # save as legacy meta_tags.csv
     tags_path = os.path.join(path_expt_dir, 'meta_tags.csv')
+    save_hparams_to_tag_csv(tags_path, hparams)
+
     tags = load_hparams_from_tags_csv(tags_path)
 
-    assert tags.batch_size == 32 and tags.hidden_dim == 1000
+    assert tags['batch_size'] == 32 and tags['hidden_dim'] == 1000
```
Comment (Borda) [inline tests/trainer/test_trainer.py:212]: you can compare complete dict so it would be sensitive if any default change

File: tests/trainer/test_trainer.py [hunk 17]
```diff
@@ -193,12 +198,37 @@ def test_loading_meta_tags(tmpdir):
     logger.log_hyperparams(hparams)
     logger.save()
 
-    # load tags
+    # load hparams
     path_expt_dir = tutils.get_data_path(logger, path_dir=tmpdir)
+    hparams_path = os.path.join(path_expt_dir, 'hparams.yaml')
```
Comment (Borda) [inline tests/trainer/test_trainer.py:203]: use here the default
```suggestion
    hparams_path = os.path.join(path_expt_dir, TensorBoardLogger.NAME_HPARAMS_FILE)
```

File: tests/trainer/test_trainer.py [hunk 18]
```diff
@@ -6,12 +7,14 @@
...
 import torch
+import yaml
 
 import tests.base.utils as tutils
 from pytorch_lightning import Callback, LightningModule
 from pytorch_lightning import Trainer
 from pytorch_lightning.callbacks import EarlyStopping, ModelCheckpoint
-from pytorch_lightning.core.lightning import load_hparams_from_tags_csv
+from pytorch_lightning import Callback
+from pytorch_lightning.core.lightning import load_hparams_from_tags_csv, load_hparams_from_yaml, save_hparams_to_yaml, save_hparams_to_tag_csv
```
Comment (Borda) [inline tests/trainer/test_trainer.py:17]:
```suggestion
from pytorch_lightning.core.lightning import load_hparams_from_tags_csv, load_hparams_from_yaml, save_hparams_to_yaml, save_hparams_to_tag_csv
from pytorch_lightning.loggers import TensorBoardLogger
```

**Comment by S-aiueo32** [comment]
how about it

**Comment by S-aiueo32** [comment]
@Borda the doc fixed 👍

File: pytorch_lightning/loggers/tensorboard.py [hunk 19]
```diff
@@ -117,15 +123,12 @@ def log_hyperparams(self, params: Union[Dict[str, Any], Namespace],
             from torch.utils.tensorboard.summary import hparams
             if metrics is None:
                 metrics = {}
-            exp, ssi, sei = hparams(sanitized_params, metrics)
+            exp, ssi, sei = hparams(params, {})
```
Comment (awaelchli) [inline pytorch_lightning/loggers/tensorboard.py:126]: was this not introduced in another PR recently? why pass in empty dict?

File: pytorch_lightning/core/lightning.py [hunk 20]
```diff
@@ -1464,19 +1466,34 @@ def __init__(self, hparams):
                 If your checkpoint saved a GPU model and you now load on CPUs
                 or a different number of GPUs, use this to map to the new setup.
                 The behaviour is the same as in :func:`torch.load`.
-            tags_csv: Optional path to a .csv file with two columns (key, value)
+            hparams_file: Optional path to a .yaml file with hierarchical structure.
                 as in this example::
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1470]:
```suggestion
            hparams_file: Optional path to a .yaml file with hierarchical structure
                as in this example::
```

File: pytorch_lightning/core/lightning.py [hunk 21]
```diff
@@ -1464,19 +1466,34 @@ def __init__(self, hparams):
...
+                        batch_size: 32
 
                 You most likely won't need this since Lightning will always save the hyperparameters
                 to the checkpoint.
                 However, if your checkpoint weights don't have the hyperparameters saved,
-                use this method to pass in a .csv file with the hparams you'd like to use.
-                These will be converted into a :class:`~argparse.Namespace` and passed into your
+                use this method to pass in a .yaml file with the hparams you'd like to use.
+                These will be converted into a :class:`~dict` and passed into your
                 :class:`LightningModule` for use.
```
Comment (awaelchli) [inline pytorch_lightning/core/lightning.py:1482]: maybe make an emphasizing note here that due to yml structure only hparams in dict structure are supported. If the model accepts a argparse Namespace, it would require the user to refactor their code.

Comment (S-aiueo32) [reply pytorch_lightning/core/lightning.py:1482]: I will add a description like:
```
If your model's `hparams` argument is :class:`~argparse.Namespace`
and .yaml file has multi-level structure, you need to refactor your model to treat
`hparams` as :class:`~dict`.
```

Is it okay?

File: pytorch_lightning/loggers/tensorboard.py (hunk 19 above)
Comment (S-aiueo32) [reply pytorch_lightning/loggers/tensorboard.py]: done 5114af2

File: pytorch_lightning/core/lightning.py (hunk 21 above)
Comment (awaelchli) [reply pytorch_lightning/core/lightning.py:1500]: saw your doc updates, nice, thanks!

File: pytorch_lightning/loggers/tensorboard.py [hunk 22]
```diff
@@ -117,15 +123,12 @@ def log_hyperparams(self, params: Union[Dict[str, Any], Namespace],
             from torch.utils.tensorboard.summary import hparams
             if metrics is None:
                 metrics = {}
-            exp, ssi, sei = hparams(sanitized_params, metrics)
+            exp, ssi, sei = hparams(params, metrics)
             writer = self.experiment._get_file_writer()
```
Comment (awaelchli) [inline pytorch_lightning/loggers/tensorboard.py:127]: Does this fix #1778? If so, I would add it to the changelog as a separate entry or perhaps make a separate PR if it is not essential to this PR.

Comment (awaelchli) [reply pytorch_lightning/loggers/tensorboard.py:127]: I mean the update above, I highlighted the wrong part

**Comment by Borda** [comment]
@S-aiueo32 mind rebase? it is the reason why Drone is failing...

**Comment by Borda** [comment]
@S-aiueo32 mind rebase? it is the reason why Drone is failing...
Could you do it today, it would be nice to have in today's release... :]

**Comment by S-aiueo32** [comment]
@Borda 
I finished my work, and now I'm working on it.
But ahhhh, I can't push my local changes after merging the remote master:
```
! [remote rejected] feature/replace-mata-tags-with-yaml -> feature/replace-mata-tags-with-yaml (refusing to allow an OAuth App to create or update workflow `.github/workflows/docker-builds.yml` without `workflow` scope)
```

Do you have any idea to resolve it?

**Comment by Borda** [comment]
> Do you have any idea to resolve it?

`git push -f`

**Comment by S-aiueo32** [comment]
doensn't work...

**Comment by Borda** [comment]
> doesn't work...

are you also on slack, can we try to resolve it there online?
or I can try to it as I suggested, squash and then rebase, ok?

**Comment by S-aiueo32** [comment]
@Borda @awaelchli @williamFalcon 
I have no counterpart to the test failure... Do you have any suggestions?

**Comment by Borda** [comment]
> I have no counterpart to the test failure... Do you have any suggestions?

this is just installation issue, not connected to this project so I just rerun the tests and it shall be fine

**Comment by S-aiueo32** [comment]
How is it going?

**Comment by Borda** [comment]
> How is it going?

I tried to rebase but I do not have permission to push it back...
As there is so many `master` merges there is no simple way to squash all 50 comics to one...
I would recommend following hack:
1. backup the actual branch - e.g. rename it with "---" at the name so you have two pointers to the same commit
2. update your master with upstream - this main repo master
3. return to this particular branch you have created the PR
4. reset to master  which should create raw diff `git reset master`
5. you shall see all changes from the PR in a stash
6. commit al these changes as a single commit
7. push it back hard to this branch `git push -f`

**Comment by Borda** [comment]
or what you just did may help, I'll check it tomorrow :]

**Comment by S-aiueo32** [comment]
@Borda @awaelchli @williamFalcon All checks pass, thanks!

**Comment by Borda** [comment]
it looks good to me, I have pinged @PyTorchLightning/core-contributors to get this done

**Comment by Borda** [comment]
@S-aiueo32 Great work! and thx for your patience :raccoon:

**Comment by Borda** [comment]
> shall we include a test for `tests/test_deprecated.py`?

completely forgot... @S-aiueo32 mind make followup or with deprecation tests
